documentation/images/plantuml/agent/state_machines/backhaul_manager_fsm.puml

```diff
@@ -0,0 +1,84 @@
+@startuml backahul_manager_fsm
+
+[*] --> INIT
+INIT --> WAIT_ENABLE
+WAIT_ENABLE --> ENABLED
+ENABLED --> MASTER_DISCOVERY : wired
+ENABLED --> WIRELESS_FSM : wireless
+state WIRELESS_FSM {
+    [*] --> INIT_HAL
+    INIT_HAL --> WPA_ATTACH
+    WPA_ATTACH --> CONNECT
+    CONNECT --> MASTER_DISCOVERY : connection success
+}
+MASTER_DISCOVERY --> OPERATIONAL
+OPERATIONAL --> RESTART : wired link status change
+RESTART --> INIT
+
+@enduml
+
+@startuml dynamic_backhaul_switch
+
+box "controller" #LightSlateGray
+participant controller
+end box
+
+box "prplmesh agent" #LightBlue
+participant backhaul_manager
+participant radio_agent
+end box
+
+box "platoform"
+participant hostapd
+participant supplicant
+end box
+
+== platform init ==
+hostapd <- : start disabled (static channel)
+activate hostapd
+deactivate hostapd
+supplicant <- : start (no network config)
+activate supplicant
+deactivate supplicant
+
+== init ==
+...
+== operational ==
+controller -> backhaul_manager : switch_backhaul(backhaul)
+activate backhaul_manager
+backhaul_manager -> backhaul_manager : restart
+note left
+m_preferred_backhaul = backhaul
+end note
+backhaul_manager -> supplicant : tear down
+note right
+disable all supplicants
+disable wired interface
+end note
+backhaul_manager -> radio_agent: disconnect
+deactivate backhaul_manager
+activate radio_agent
+radio_agent -> radio_agent : restart
+radio_agent -> hostapd : tear down
```

Comment:
That's not good... When we switch backhaul, we *must not* tear down the APs! In fact, in a steering scenario, ideally we should only remove the old backhaul when the new one is fully up and usable.

Reply:
Switching from one backhaul band to another requires changing the fronthaul channel - which means we need to either manually steer the clients before we do the switch or tear down. I think that for M1/M2 tear down is fine, which is why I chose it. I can update the design to say that in M3 we will do the steering for stations before we steer the backhaul.

Reply:
As an example, consider a dual-band repeater connected with 5G backhaul.
The 2.4G far AP is on channel 6, while the repeater fronthaul 2.4G is on channel 1.
The controller sends a backhaul steering request to the far 2.4G AP, so the clients connected to the repeater's 2.4G front AP needs to switch channel.
The bSTA however needs to do scan, and since there is no background scan, it has to tear down the front AP. This is how I understood this, please correct me if I'm wrong.

Reply:
OK, for M2 doing teardown is fine.

If possible, it would be nice to make sure that we only tear down the radio on which the backhaul is active. But nice-to-have for M2.

Note that in a wireless-backhaul-only scenario, you would typically configure all APs on the same channel for exactly this reason. Let's say the gateway has a bAP on channel 36. Then repeater1 that is connected to it must have its bSTA on channel 36 as well. If it operates a bAP on the same radio, it will be on channel 36 too. Therefore, repeater2 that connects to repeater1 must have its bSTA on channel 36 as well. And so on. In other words, in practice a channel switch is only required if there's at least one wired link in the network.

For the scan, same thing: if it stays on the same channel, it can do the scan while operating the AP. Well, if the driver supports that, of course...

Reply:
Agreed.